services/DeviceInformation/source/DeviceInformationService.cpp

```diff
@@ -0,0 +1,127 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2020 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ble-service-device-information/DeviceInformationService.h"
+
+#if BLE_FEATURE_GATT_SERVER
+
+ble_error_t DeviceInformationService::add_service(
+    BLE &ble,
+    const char *manufacturers_name,
+    const char *model_number,
+    const char *serial_number,
+    const char *hardware_revision,
+    const char *firmware_revision,
+    const char *software_revision,
+    const SystemId *system_id,
+    const RegulatoryCertDataList *cert_data_list,
+    const PnPID *pnp_id
+) {
+    /* all characteristics are optional so we need to build an array that contains the requested ones */
+    size_t param_index = 0;
+
+    const char* param_strings[6];
+    UUID param_string_uuids[6];
+
+    if (manufacturers_name) {
+        param_strings[param_index] = manufacturers_name;
+        param_string_uuids[param_index++] = GattCharacteristic::UUID_MANUFACTURER_NAME_STRING_CHAR;
+    }
+    if (model_number) {
+        param_strings[param_index] = model_number;
+        param_string_uuids[param_index++] = GattCharacteristic::UUID_MODEL_NUMBER_STRING_CHAR;
+    }
+    if (serial_number) {
+        param_strings[param_index] = serial_number;
+        param_string_uuids[param_index++] = GattCharacteristic::UUID_SERIAL_NUMBER_STRING_CHAR;
+    }
+    if (hardware_revision) {
+        param_strings[param_index] = hardware_revision;
+        param_string_uuids[param_index++] = GattCharacteristic::UUID_HARDWARE_REVISION_STRING_CHAR;
+    }
+    if (firmware_revision) {
+        param_strings[param_index] = firmware_revision;
+        param_string_uuids[param_index++] = GattCharacteristic::UUID_FIRMWARE_REVISION_STRING_CHAR;
+    }
+    if (software_revision) {
+        param_strings[param_index] = software_revision;
+        param_string_uuids[param_index++] = GattCharacteristic::UUID_SOFTWARE_REVISION_STRING_CHAR;
+    }
+
+    GattCharacteristic* param_chars[9];
+
+    for (int i = 0; i < param_index; i++) {
+        param_chars[i] = new GattCharacteristic(
+            param_string_uuids[i],
+            (uint8_t *)param_strings[i],
+            strlen(param_strings[i]), /* Min length */
```

Comment:
Is there a maximal lenght defined in the specification ?

Reply:
no